crates/polars-plan/src/dsl/function_expr/range/utils.rs

```diff
@@ -19,3 +22,86 @@ pub(super) fn ensure_range_bounds_contain_exactly_one_value(
     );
     Ok(())
 }
+
+/// Create a ranges column from the given start/end columns and a range function.
+pub(super) fn ranges_impl_broadcast<T, U, F>(
+    start: &ChunkedArray<T>,
+    end: &ChunkedArray<T>,
+    range_impl: F,
+    builder: &mut ListPrimitiveChunkedBuilder<U>,
+) -> PolarsResult<Series>
+where
+    T: PolarsIntegerType,
+    U: PolarsIntegerType,
+    F: Fn(T::Native, T::Native, &mut ListPrimitiveChunkedBuilder<U>) -> PolarsResult<()>,
+{
+    match (start.len(), end.len()) {
+        (len_start, len_end) if len_start == len_end => {
+            build_ranges::<T, U, F>(start.into_iter(), end.into_iter(), range_impl, builder)?;
+        },
+        (1, len_end) => {
+            let start_scalar = start.get(0);
+            match start_scalar {
+                Some(start) => build_ranges::<T, U, F>(
+                    repeat(Some(start)),
+                    end.into_iter(),
+                    range_impl,
+                    builder,
+                )?,
+                None => build_nulls(builder, len_end),
+            }
+        },
+        (len_start, 1) => {
+            let end_scalar = end.get(0);
+            match end_scalar {
+                Some(end) => build_ranges::<T, U, F>(
+                    start.into_iter(),
+                    repeat(Some(end)),
+                    range_impl,
+                    builder,
+                )?,
+                None => build_nulls(builder, len_start),
+            }
+        },
+        (len_start, len_end) => {
+            polars_bail!(
+                ComputeError:
+                "lengths of `start` ({}) and `end` ({}) do not match",
+                len_start, len_end
+            )
+        },
+    };
+    let out = builder.finish().into_series();
+    Ok(out)
+}
+
+/// Iterate over a start and end column and create a range for each entry.
+fn build_ranges<T, U, F>(
+    start: impl Iterator<Item = Option<T::Native>>,
```

Comment:
nit: This is entirely correct, but we normally use the generic type here in the. E.g. `I` and `J` and write the bound either inline in the function definition or in a `where` clause.

This makes it a bit easier to extend the trait bounds when we need to add extra in the future. And it is possible to call the trait methods/items directly. E.g. `I::Item` or `I::next`.

Reply:
Ok, makes sense. I had it like that but a function with 5 generics felt a bit unwieldy. Updated!